<h3>1. Analyze</h3>

In [1]:
import pandas as pd

we are accessing the csv file through the first line of code then we are displaying the first part of the data through the second line

In [3]:
dataset=pd.read_csv("kc_house_data.csv")
dataset.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

Now let's start the analyse based on the headers that we have:


In [4]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

First we see that this data is representing the prices and the description of appartements or buildings in a certain area.We can see that the dominant data type in this data is int64.Then we can remark that each appartment or building has a specific id.

In [58]:
price_condition=dataset[["price","condition","bedrooms"]]
floor_basement_room=dataset[["floors","sqft_basement","bedrooms"]]
loc=dataset[["lat","long"]]


max_condition=dataset.condition.max()

#this .sort_values helps sort the values by price and in a descending way so we can check if the values are related to each other
sorted_dataset_price=price_condition.sort_values(by='price',ascending=False)
sorted_dataset_floors=floor_basement_room.sort_values(by="floors",ascending=False)


floors_basement=dataset[(dataset.floors >=1) & (dataset.sqft_basement >0)]
floors_bedrooms=dataset[(dataset.floors >=1) & (dataset.bedrooms==0)]
wrong_val=dataset[(dataset.bedrooms==0) | (dataset.sqft_living==0) | (dataset.zipcode ==0) ]


min_floor=dataset.floors.min()
min_basement=dataset.sqft_basement.min()
max_bedrooms=dataset.bedrooms.max()

#idmax() is used to retrieve the index so when I am using value_counts() the index is the value itself
most_building_year=dataset["yr_built"].value_counts().idxmax()
loc_of_mostBuild=loc.value_counts()


print(wrong_val[["bedrooms","zipcode"]])
#print(dataset['yr_built'].value_counts())
#print(floors_bedrooms[["floors","bedrooms"]])
#print(sorted_dataset_floors)
# print(sorted_dataset.head(10))
print("best condition:",max_condition)
print("biggest number of bedrooms: ",max_bedrooms)
print("the smallest floor is: ",min_floor)
print("the smallest squrft of a basement: ",min_basement)
print("the largest bedrooms are ",max_bedrooms)
print("the year that has the most building in it is: ",most_building_year)
print("the location of most of the houses: ",loc_of_mostBuild.idxmax())

Empty DataFrame
Columns: [bedrooms, zipcode]
Index: []
best condition: 5
biggest number of bedrooms:  33
the smallest floor is:  1.0
the smallest squrft of a basement:  0
the largest bedrooms are  33
the year that has the most building in it is:  2014
the location of most of the houses:  (47.6599, -122.146)


Secondly, we can find a lot of illogic values in this dataset.To start on this issue we are checking the normally related topics.For example we can see find that the most expensive price is not related to the best condition,and also the number of bedrooms is not related to the maximum price.In addition, Furthermore we can see that some buildings have 3.5 floors with 0 bedrooms which doesn't make sense it could be a storage facility but we would need more details on the data given.

Thirdly we can detect that the average of these buildings are contemporary and built in the year 2014.The length of the value count is 20832 which is not a small number compared to the entries of the data;so we can conclude that the buildings available are distributed in an equal way because the most building in the same area are 4 and the number of entry is 21613.

<h3><b>2.Testing</b></h3> 

<h5><b>a. Building a dataframe from a dictionnary</b> </h5>

In [63]:
#this is testing if this is the right way to set up a data frame
#!!!!the length if the arrays must be the same!!!!
test={
    "price": [1,5,7],
    "width": [24.5,45.0,20.7]
}

#we need to convert the dict to a data frame =>
df=pd.DataFrame(test)
print(df)

   price  width
0      1   24.5
1      5   45.0
2      7   20.7


<h5><b>b. If column contains integers and there is a string in it</b></h5>

In [65]:
test2={
    "price": [1,'asfih',5],
    "width": [1,3,5]
}

df2=pd.DataFrame(test2)
print(df2.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   price   3 non-null      object
 1   width   3 non-null      int64 
dtypes: int64(1), object(1)
memory usage: 180.0+ bytes
None


So when we have a string or char inside of a integer column the column type will be converted to an <font color="red">Object</font>

<h5><b>c. If column of integers have floats</b></h5>

In [66]:
test2={
    "price": [1,2,5],
    "width": [1,3.5,5]
}

df2=pd.DataFrame(test2)
print(df2.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   price   3 non-null      int64  
 1   width   3 non-null      float64
dtypes: float64(1), int64(1)
memory usage: 180.0 bytes
None


So when we have a float inside of an integer column or vise versa the column type will be converted to an <font color="red">Float (64 bit)</font>

<h5><b>d. If column of booleans have a string</b></h5>

In [70]:
test3={
    "price": [True,False,"s"],
    "width": [1,3.5,True]
}

df3=pd.DataFrame(test3)
print(df3.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   price   3 non-null      object
 1   width   3 non-null      object
dtypes: object(2)
memory usage: 180.0+ bytes
None


So when we have a String or an integer or float inside of a boolean column or vise versa the column type will be converted to an <font color="red">Object</font>

<h4><font color="red">Conclusion</font></h4>

If we have an integer,float,string or char inside of a String column or vise versa : <font color='red'>=>Object</font>


If we have an integer inside of a float column or vice versa it will convert to the bigger: <font color='red'>=>Float(64)</font>